kahypar/partition/preprocessing/louvain.h

```diff
@@ -25,229 +25,227 @@
-                if(std::abs(cur_quality-quality.quality()) > EPS) {
-                    LOGVAR(cur_quality);
-                    LOGVAR(quality.quality());
-                    return false;
-                }
-                return true;
-            }(),"Quality of the contracted graph is not equal with quality of its corresponding uncontracted graph!");
-            
-            old_quality = cur_quality;
-            HighResClockTimepoint start = std::chrono::high_resolution_clock::now();
-            cur_quality = louvain_pass(graph_stack[cur_idx],quality);
-            HighResClockTimepoint end = std::chrono::high_resolution_clock::now();
-            std::chrono::duration<double> elapsed_seconds = end - start;
-            LOG("Louvain-Pass #" << iteration << " Time: " << elapsed_seconds.count() << "s");
-            improvement = cur_quality - old_quality > _config.preprocessing.louvain_community_detection.min_eps_improvement || max_iterations == 2;
-            
-            LOG("Louvain-Pass #" << iteration << " improve quality from " << old_quality << " to " << cur_quality);
-            
-            if(improvement) {
-                cur_quality = quality.quality();
-                LOG("Starting Contraction of communities...");
-                start = std::chrono::high_resolution_clock::now();
-                auto contraction = graph_stack[cur_idx++].contractCluster();
-                end = std::chrono::high_resolution_clock::now();
-                elapsed_seconds = end - start;
-                LOG("Contraction Time: " << elapsed_seconds.count() << "s");
-                graph_stack.push_back(std::move(contraction.first));
-                mapping_stack.push_back(std::move(contraction.second));
-                LOG("Current number of communities: " << graph_stack[cur_idx].numNodes());
-            }
-            
-            LOG("");
-            
-        } while(improvement && iteration < max_iterations);
-        
-        ASSERT((mapping_stack.size() + 1) == graph_stack.size(), "Unequality between graph and mapping stack!");
-        while(!mapping_stack.empty()) {
-            assignClusterToNextLevelFinerGraph(graph_stack[cur_idx-1],graph_stack[cur_idx],mapping_stack[cur_idx-1]);
-            graph_stack.pop_back();
-            mapping_stack.pop_back();
-            cur_idx--;
-        }
-        
-        for(NodeID node : _graph.nodes()) { 
-            _graph.setClusterID(node,graph_stack[0].clusterID(node));    
-        }    
-        
-        
-        return cur_quality;
-        
-    }
-    
-    Graph getGraph() {
-        return _graph;
-    }
-    
-    ClusterID clusterID(const HypernodeID hn) const {
-        return _graph.hypernodeClusterID(hn);
+ public:
+  Louvain(const Hypergraph& hypergraph,
+          const Configuration& config) :
+    _graph_hierarchy(),
+    _config(config) {
+    _graph_hierarchy.emplace_back(hypergraph, config);
+  }
+
+  Louvain(const std::vector<NodeID>& adj_array,
```

Comment:
No, it's also needed in the karate club test case, since you build the adjacency structure there.